Add a documentation page to explain how to implement undo/redo in a third-party editor (#54546)
WordPress/gutenberg: platform-docs/docs/basic-concepts/undo-redo.md

---
sidebar_position: 3
---

# Undo Redo

## Undo/Redo using `useStateWithHistory`

> By default the block editor doesn't enable undo/redo. This is because in most scenarios, the block editor is rendered as part of a larger application that already has its own undo/redo functionality. However, to help you implement undo/redo in your application, the block editor provides a set of APIs that you can use.

Reviewer: ```suggestion
By default, the block editor doesn't enable undo/redo. This is because in most scenarios, the block editor is rendered as part of a larger application that already has its own undo/redo functionality. However, to help you implement undo/redo in your application, the block editor provides a set of APIs that you can use.
```

The simplest approach is to rely on the `useStateWithHistory` hook provided by the `@wordpress/compose` package. This hook is a wrapper around the `useState` hook that adds undo/redo functionality to the state. 

First, make sure you add the `@wordpress/compose` package to your dependencies, then use the hook like so:

> ```js

Reviewer 2: ```suggestion
```jsx
```

jsx also highlights components and props, so I think it's easier to understand.

| js | jsx |
|--------|--------|
| ![js](https://github.com/WordPress/gutenberg/assets/54422211/2741ffcc-0e0a-4204-960e-08b2ec5abcf9) | ![jsx](https://github.com/WordPress/gutenberg/assets/54422211/24937493-a439-4ae4-8a9d-a492e532c765) |

import { useStateWithHistory } from '@wordpress/compose';
import { createRoot, createElement, useState } from "@wordpress/element";
import {
  BlockEditorProvider,
  BlockCanvas,
} from "@wordpress/block-editor";

> 

Reviewer 2: Do you want to omit things like `@wordpress/components` and `@wordpress/icons` packages that are not the focus of this page?

Author: I just forgot about the imports :) but yeah I might change the code to avoid these dependencies and focus on the purpose of the page.

                    disabled={ ! hasUndo }
                    icon={ undoIcon }
                    label="Undo"
                />
                <Button
                    onClick={ redo }
                    disabled={ ! hasRedo }
                    icon={ redoIcon }
                    label="Redo"
                />
            </div>
            <BlockCanvas />
        </BlockEditorProvider>
	);
}
```

The `useStateWithHistory` hook returns an object with the following properties:

- `value`: the current value of the state.
- `setValue`: a function that can be used to update the state.
- `hasUndo`: a boolean indicating whether there are any actions that can be undone.
- `hasRedo`: a boolean indicating whether there are any actions that can be redone.
- `undo`: a function that can be used to undo the last action.
- `redo`: a function that can be used to redo the last action.

Notice that in addition to the `blocks` property, the `value` object also tracks a `selection` property. This property is used to store and control the cursor position within the block editor. This allows the editor to restore the right position when undoing or redoing changes.

## Going Further...

> Often times, editors allows to track changes accross multiple objects and properties. For instance a basic writing experience might allow editing the title of a post in a normal input, and the content of the post in the block editor. Or your editor might allow edits to related objects like categories, tags... In these cases, you might want to implement undo/redo functionality that can be used to track changes across all of these objects and properties.

Reviewer: ```suggestion
Oftentimes, editors allow to track changes across multiple objects and properties. For instance, a basic writing experience might allow editing the title of a post in a normal input and the content of the post in the block editor. Or your editor might allow edits to related objects like categories, tags, etc... In these cases, you might want to implement undo/redo functionality that can be used to track changes across all of these objects and properties.
```
